File: 442_FindAllDuplicatesInAnArray.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDuplicates = function(nums) {
    var l = nums.length, r = [], arr = new Array(l);
    for (var i = 0; i < l; ++i) {
        arr[i] = -1 * (i + 1);
    }
    for (var i = 0; i < l; ++i) {
        arr[nums[i] - 1] += nums[i];
    }
    for (var i = 0; i < l; ++i) {
        if (arr[i] > 0) {
            r.push(i + 1);
        }
    }
    return r;
};

File: 442.FindAllDuplicatesinanArray.java
import java.util.ArrayList;

class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        List<Integer> result = new ArrayList<Integer>();
        for (int i = 0; i < nums.length; i += 1) {
            int index = Math.abs(nums[i]) - 1;
            if (nums[index] < 0) {
                result.add(index + 1);
            }
            nums[index] = -nums[index];
        }
        return result;
    }
}